beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs

```diff
@@ -793,4 +894,109 @@ impl<T: BeaconChainTypes> ReprocessQueue<T> {
             self.lc_updates_delay_queue.len() as i64,
         );
     }
+
+    fn recompute_next_backfill_batch_event(&mut self) {
+        // only recompute the `next_backfill_batch_event` if there are backfill batches in the queue
+        if !self.queued_backfill_batches.is_empty() {
+            self.next_backfill_batch_event = Some(Box::pin(tokio::time::sleep(
+                ReprocessQueue::<T>::duration_until_next_backfill_batch_event(&self.slot_clock),
+            )));
+        } else {
+            self.next_backfill_batch_event = None
+        }
+    }
+
+    /// Returns duration until the next scheduled processing time. The schedule ensure that backfill
+    /// processing is done in windows of time that aren't critical
+    fn duration_until_next_backfill_batch_event(slot_clock: &T::SlotClock) -> Duration {
+        let slot_duration = slot_clock.slot_duration();
+        slot_clock
+            .seconds_from_current_slot_start()
+            .and_then(|duration_from_slot_start| {
+                BACKFILL_SCHEDULE_IN_SLOT
+                    .iter()
+                    .map(|percentage_in_slot| {
+                        // convert % to seconds from slot start
+                        (percentage_in_slot * slot_duration.as_secs_f32()) as u64
+                    })
+                    .find_or_first(|event_time_from_slot_start| {
+                        *event_time_from_slot_start > duration_from_slot_start.as_secs()
+                    })
+                    .map(|next_event_time_secs| {
+                        if duration_from_slot_start.as_secs() >= next_event_time_secs {
+                            // event is in the next slot, add duration to next slot
+                            let duration_to_next_slot = slot_duration - duration_from_slot_start;
+                            duration_to_next_slot + Duration::from_secs(next_event_time_secs)
+                        } else {
+                            Duration::from_secs(
+                                next_event_time_secs - duration_from_slot_start.as_secs(),
+                            )
+                        }
+                    })
+            })
```

Comment:
I notice that there's a loss of precision here from two factors:

1. Casting float to int (this will [round to zero](https://doc.rust-lang.org/reference/expressions/operator-expr.html#semantics)).
1. Casting `Duration` to seconds.

AFAICT, the way that you have the two interacting together will work without errors. However I did want to consider what it would be like *without* rounding.

Because of (1), we're rounding down the `BACKFILL_SCHEDULE_IN_SLOT` fractions to zero. This results in the following schedules:

Mainnet (12s slots):

- 50% is 6s
- 60% is 7s (7.2s without rounding)
- 80% is 9s (9.6s without rounding)

Gnosis (5s slots):

- 50% is 2 (or 2.5s with rounding).
- 60% is always 3s
- 80% is always 4s

I wonder if we're better off with or without rounding 🤔 Is this something you've considered? I feel like the implementation is simpler without rounding and it gives us some more control in when we start the tasks. Without rounding we also have more freedom to jig around with the mainnet schedule without accidentally landing two schedule periods on the same second for Gnosis.

Here's an implementation *without* rounding that I made just to make sure it's sane (not this implementation uses the non-float schedule I suggested earlier):


```suggestion
                BACKFILL_SCHEDULE_IN_SLOT
                    .into_iter()
                    // Convert fractions to seconds from slot start.
                    .map(|(multiplier, divisor)| (slot_duration / divisor) * multiplier)
                    .find_or_first(|&event_duration_from_slot_start| {
                        event_duration_from_slot_start > duration_from_slot_start
                    })
                    .map(|next_event_time| {
                        if duration_from_slot_start >= next_event_time {
                            // event is in the next slot, add duration to next slot
                            let duration_to_next_slot = slot_duration - duration_from_slot_start;
                            duration_to_next_slot + next_event_time
                        } else {
                            next_event_time - duration_from_slot_start
                        }
                    })
            })
```

Reply:
Thanks for the feedback! Yep, I've considered with or without rounding but didn't put in too much thought. My initial implementation was to process at 6, 7, 10 seconds without the fractions, but moved to fractions later in order to support different slot time configurations, and I originally thought rounding it to integer seconds would be simpler, not from an implementation perspective but from users perspective. However I'm now convinced that we're better off without rounding, since backfill processing is pretty much transparent to the users and rounding doesn't really help in any way, but adds complexity. I'll update the implementation to what you've suggested, I think it's definitely much cleaner!

Reply:
also I think there's less potential issues without rounding, we might end up processing less times in a slot depending on the configuration and slot times, so agree not rounding gives us more control 👍

Reply:
Updated the code and test. The failing test revealed that I can't use `slot_clock.seconds_from_current_slot_start` if we're not rounding, so I've added a new `millis_from_current_slot_start` function.